File: Editorial/1301-1400/1320_Minimum_Distance_to_Type_a_Word_Using_Two_Fingers/C/2_Dynamic_Programming_Space_Optimization.c
int getDistance(int p, int q) {
    int x1 = p / 6, y1 = p % 6;
    int x2 = q / 6, y2 = q % 6;
    return abs(x1 - x2) + abs(y1 - y2);
}

int minimumDistance(char* word) {
    int n = strlen(word);
    int** dp = (int**)malloc(n * sizeof(int*));
    for (int i = 0; i < n; i++) {
        dp[i] = (int*)malloc(26 * sizeof(int));
        for (int j = 0; j < 26; j++) {
            dp[i][j] = INT_MAX / 2;
        }
    }
    for (int j = 0; j < 26; j++) {
        dp[0][j] = 0;
    }

    for (int i = 1; i < n; i++) {
        int cur = word[i] - 'A';
        int prev = word[i - 1] - 'A';
        int d = getDistance(prev, cur);

        for (int j = 0; j < 26; j++) {
            dp[i][j] = fmin(dp[i][j], dp[i - 1][j] + d);
            if (prev == j) {
                for (int k = 0; k < 26; k++) {
                    int d0 = getDistance(k, cur);
                    dp[i][j] = fmin(dp[i][j], dp[i - 1][k] + d0);
                }
            }
        }
    }

    int ans = INT_MAX / 2;
    for (int j = 0; j < 26; j++) {
        if (ans > dp[n - 1][j]) {
            ans = dp[n - 1][j];
        }
    }
    for (int i = 0; i < n; i++) {
        free(dp[i]);
    }
    free(dp);

    return ans;
}

File: Editorial/1301-1400/1320_Minimum_Distance_to_Type_a_Word_Using_Two_Fingers/Go/2_Dynamic_Programming_Space_Optimization.go
func getDistance(p, q int) int {
	x1, y1 := p/6, p%6
	x2, y2 := q/6, q%6
	return abs(x1-x2) + abs(y1-y2)
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func minimumDistance(word string) int {
	n := len(word)
	dp := make([][]int, n)
	for i := range dp {
		dp[i] = make([]int, 26)
		for j := range dp[i] {
			dp[i][j] = 1 << 30
		}
	}
	for j := 0; j < 26; j++ {
		dp[0][j] = 0
	}

	for i := 1; i < n; i++ {
		cur := int(word[i] - 'A')
		prev := int(word[i-1] - 'A')
		d := getDistance(prev, cur)

		for j := 0; j < 26; j++ {
			dp[i][j] = min(dp[i][j], dp[i-1][j]+d)
			if prev == j {
				for k := 0; k < 26; k++ {
					d0 := getDistance(k, cur)
					dp[i][j] = min(dp[i][j], dp[i-1][k]+d0)
				}
			}
		}
	}

	ans := 1 << 30
	for j := 0; j < 26; j++ {
		ans = min(ans, dp[n-1][j])
	}
	return ans
}
